pipeline/hard-reset.sh

```diff
@@ -0,0 +1,343 @@
+progname="${0##*/}"
+case "${0}" in
+*/*) progdir="${0%/*}";;
+*) progdir=".";;
+esac
+
+. "${progdir}/msg.sh"
+. "${progdir}/usage.sh"
+. "${progdir}/util.sh"
+. "${progdir}/common.sh"
+
+print_usage() {
+   cat <<- ENDEND
+
+      usage: ${progname} -p <profile> [options] shard command
+
+      options:
+      -p          mandatory: profile of network to run on
+                  (support only stn/os/ps)
+
+      -o outdir   use the given output directory
+                  (default: the hard-reset/YYYY-MM-DDTHH:MM:SSZ subdir logs/profile)
+
+      -y          say yes to cmd confirmation
+
+      shard       the shard number, such as 0-3 ('all' means all shards)
+
+      commands:
+
+      reset       do the hard reset, everything
+
+      network     do the network hard reset only
+      explorer    reset explorer
+      dashboard   reset staking dashboard
+      sentry      reset/launch sentry nodes
+      regression  start regression test
+            
+ENDEND
+}
+
+######### variables #########
+unset -v net_profile force_yes network dbprefix
+force_yes=false
+net_profile=
+unset -v OPTIND OPTARG opt
+OPTIND=1
+while getopts :o:p:y opt
+do
+   case "${opt}" in
+   '?') usage "unrecognized option -${OPTARG}";;
+   ':') usage "missing argument for -${OPTARG}";;
+   o) outdir="${OPTARG}";;
+   p) net_profile="${OPTARG}"
+      BENCHMARK_FILE="$CONFIG_DIR/benchmark-${net_profile}.json"
+      [ ! -e "$BENCHMARK_FILE" ] && errexit "can't find benchmark config file : $BENCHMARK_FILE"
+      export HMY_PROFILE=${net_profile}
+      ;;
+   y) force_yes=true;;
+   *) err 70 "unhandled option -${OPTARG}";;
+   esac
+done
+shift $(( OPTIND - 1 ))
+
+#### cmdline checking ####1
+unset -v shard cmd
+shard="${1-}"
+shift 1 2> /dev/null || usage "missing shard argument"
+cmd="${1-}"
+shift 1 2> /dev/null || usage "missing command argument"
+
+if [ -z "${net_profile}" ] ;then
+   msg "profile not set, exiting..."
+   msg "please use -p option to specify profile"
+   exit 1
+fi
+logdir="logs/${net_profile}"
+msg "profile: ${net_profile}"
+msg "execute: ${cmd}"
+if [ "${force_yes}" = false ] ;then
+  printf "[Y]/n > "
+  read -r yn
+  if [ "${yn}" != "Y" ] ;then
+     exit
+  fi
+fi
+
+######### functions ###########
+hard_reset_shard
+{
+   shard=$1
+   msg "cat $logdir/shard${shard}.txt | xargs -i{} -P50 bash -c \"./restart_node.sh -U -X -D -y -d logs/${net_profile} -p ${net_profile} -t 0 -r 0 -R 0 {}\""
+   cat "$logdir/shard${shard}.txt" | xargs -i{} -P50 bash -c "./restart_node.sh -U -X -D -y -d logs/${net_profile} -p ${net_profile} -t 0 -r 0 -R 0 {}"
+}
+
+check_consensus
+{
+   shard=$1
+   msg "./run_on_shard.sh -p ${net_profile} -T $shard 'tac /home/tmp_log/*/zeorlog*.log | grep -m 1 HOORAY'"
+}
+
+check_version
+{
+   shard=$1
+   msg "./run_on_shard.sh -p ${net_profile} -T $shard './harmony -version'"
+}
+
+restart_watchdog
+{
+   msg "./restart_watchdog.sh -a restart -s $network"
+   ./restart_watchdog.sh -a restart -s "$network"
```

Comment:
Are the IP lists assumed to stay the same?
If not, use `-u` to pull latest nodedb before restart.

Reply:
for hard-reset, we assume the same ip, not a fresh new network.